src/jdk.crypto.ec/share/classes/sun/security/ec/ECOperations.java

```diff
@@ -486,5 +409,254 @@ public boolean checkOrder(ECPoint point) {
+                        int index = (posit(s, pos + 192) << 3) |
+                                    (posit(s, pos + 128) << 2) |
+                                    (posit(s, pos +  64) << 1) |
+                                     posit(s, pos);
+
+                        lookup(P256.points[j], index, r);
+                        Secp256R1Ops.instance.setSum(d, r, t0, t1, t2, t3, t4);
+                    }
+                }
+
+                return d;
+            }
+
+            private static int posit(byte[] k, int i) {
+                return (k[i >> 3] >> (i & 0x07)) & 0x01;
+            }
+        }
+
+        final class P256 {
+            private static final ProjectivePoint.Immutable[][] points;
+
+            // Generate the pre-computed tables.  This block may be replaced
+            // with hard-coded tables in order to speed up the class loading.
+            static {
+                points = new ProjectivePoint.Immutable[4][16];
+                BigInteger[] factors = new BigInteger[] {
+                        BigInteger.ONE,
+                        BigInteger.TWO.pow(64),
+                        BigInteger.TWO.pow(128),
+                        BigInteger.TWO.pow(192)
+                };
+
+                BigInteger[] base = new BigInteger[16];
+                base[0] = BigInteger.ZERO;
+                base[1] = BigInteger.ONE;
+                base[2] = factors[1];
+                for (int i = 3; i < 16; i++) {
+                    base[i] = BigInteger.ZERO;
+                    for (int k = 0; k < 4; k++) {
+                        if (((i >>> k) & 0x01) != 0) {
+                            base[i] = base[i].add(factors[k]);
+                        }
+                    }
+                }
+
+                for (int d = 0; d < 4; d++) {
+                    for (int w = 0; w < 16; w++) {
+                        BigInteger bi = base[w];
+                        if (d != 0) {
+                            bi = bi.multiply(BigInteger.TWO.pow(d * 16));
+                        }
+                        if (w == 0) {
+                            points[d][0] = new ProjectivePoint.Immutable(
+                                    Secp256R1.zero.fixed(),
+                                    Secp256R1.zero.getField().get1().fixed(),
+                                    Secp256R1.zero.fixed());
+                        } else {
+                            PointMultiplier multiplier = new Default(
+                                    Secp256R1Ops.instance,
+                                    AffinePoint.fromECPoint(Secp256R1.point,
+                                            Secp256R1.zero.getField()));
+                            byte[] s = bi.toByteArray();
+                            ArrayUtil.reverse(s);
+                            ProjectivePoint.Mutable m =
+                                    multiplier.pointMultiply(s);
+                            points[d][w] = m.setValue(m.asAffine()).fixed();
```

Comment:
Did you try using AffinePoints instead of ProjectivePoints? Sum with AffinePoints is supposed to be a bit faster, wonder if the improvement is worthwhile.

Also, unless you switch to affine points, just use `m.fixed()`.

Reply:
ah... right. The neutral point, aka point at infinity (0:1:0) cannot be represented in affine space. Thanks for checking.